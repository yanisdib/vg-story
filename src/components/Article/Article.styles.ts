import styled from 'styled-components';


interface StyledViewProps {
    background: string;
}


export const View = styled.div<StyledViewProps>`
display: flex;
flex-direction: row;
min-height: 100vh;
background: ${(
    { background = '#000' }: StyledViewProps
) => `linear-gradient(180deg, ${background} 70vh, #FFF 0%)`};
`;

export const Wrapper = styled.div`
display: flex;
flex-direction: column;
width: 100%;
height: 100%;
`;

export const Content = styled.div`
display: flex;
flex-direction: column;
width: 100%;
height: auto;
`;

export const Details = styled.div`

    padding: calc(var(--ui-padding-y) / 2) 0;
    line-height: var(--ui-title-lh-xl);

    a {
        color: #000;
        text-decoration: none #000;
            
        &:hover {
            text-decoration: underline #000;
        }
    }
`;

export const FrontImage = styled.div`

    display: flex;
    flex-direction: row;
    justify-content: center;
    width: 100%;
    max-height: 950px;

    img {
        display: flex;
        width: auto;
        height: auto;
        max-width: 100%;
        max-height: 100%;
        object-fit: cover;
        object-position: center;
    }
`;

export const Main = styled.main`

    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    padding: calc(var(--ui-padding-y) / 2) 0;
`;

export const Information = styled.div`

    display: flex;
    flex-direction: column;
    flex: 30%;

    span {
        font-family: var(--ui-title-font);
        font-size: 1.1em;
        padding: calc(var(--ui-padding-y) / 10) 0;
        color: #a5a6ac;
    }
`;

export const Body = styled.div`

    display: flex;
    flex-direction: column;
    flex: 70%;
`;