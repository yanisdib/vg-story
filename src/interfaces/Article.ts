import {
    Comment,
    ImageAttributes,
    Topic
} from '.';


/**
 * Article is an interface which describes properties
 * an article should be created with
 */
export interface Article {
    author: string;
    backgroundHex?: string;
    body?: string;
    comments?: Comment[];
    createdAt: number;
    editedAt?: number;
    frontImage: ImageAttributes;
    imageGallery?: {
        id: string,
        images: ImageAttributes[]
    }[];
    id: string;
    isFeatured?: boolean;
    permalink: string;
    title: string;
    topics: Topic[];
}