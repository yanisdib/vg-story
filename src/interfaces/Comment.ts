/**
 * Comment is an interface which describes properties
 * an article's comment should be created with
 */
export interface Comment {
    id: string;
    body: string;
    createdAt: number;
}