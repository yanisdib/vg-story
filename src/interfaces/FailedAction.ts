import { Action } from '@reduxjs/toolkit';


/**
 * FailedAction is an interface which describes
 * properties returned by the payload of
 * an action that failed
 */
export interface FailedAction extends Action {
    error: Error;
}