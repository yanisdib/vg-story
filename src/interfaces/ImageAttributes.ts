/**
 * Attributes an image object should be declared with
 */
export interface ImageAttributes {
    alt?: string;
    src: string;
    title?: string;
}