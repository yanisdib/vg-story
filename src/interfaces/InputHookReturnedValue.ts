export type InputHookReturnedValue<T> = [
    T,
    (event: React.ChangeEvent<HTMLInputElement>) => void
]
