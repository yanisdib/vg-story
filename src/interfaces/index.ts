export type { Article } from './Article';
export type { Comment } from './Comment';
export type { ImageAttributes } from './ImageAttributes';
export type { Topic } from './Topic';

export type { FailedAction } from './FailedAction';
export type { StoreState } from './StoreState';
