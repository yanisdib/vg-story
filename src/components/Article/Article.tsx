import moment from 'moment';
import React from 'react';
import { Link } from 'react-router-dom';

import { Article as IArticle } from '../../interfaces';

import Container from '../Container/Container';

import {
    Body,
    Content,
    Details,
    FrontImage,
    Information,
    Main,
    View,
    Wrapper
} from './Article.styles';


interface ArticleProps {
    currentArticle: IArticle;
}


function Article({ currentArticle }: ArticleProps): JSX.Element {

    const {
        author,
        backgroundHex = '#000',
        body,
        comments,
        createdAt,
        frontImage,
        imageGallery,
        title,
        topics
    }: IArticle = currentArticle;

    return (
        <View background={backgroundHex}>
            <Wrapper>
                <Container>
                    <Content>
                        <Details>
                            <h5>
                                {topics
                                    .map<React.ReactNode>(topic => (
                                        <Link key={topic.id} to={`/browse/${topic.permalink}`}>
                                            {topic.name}
                                        </Link>
                                    ))
                                    .reduce((previous, current) => [
                                        previous,
                                        ' — ',
                                        current
                                    ])
                                }
                            </h5>
                            <h2>{title}</h2>
                            <h5>
                                {`Written by ${author} — ${frontImage?.title ?? ''}`}
                            </h5>
                        </Details>
                        <FrontImage>
                            <img
                                alt={frontImage?.alt ?? title}
                                src={frontImage.src}
                                title={frontImage?.title ?? title}
                            />
                        </FrontImage>
                        <Main>
                            <Information>
                                <span>Published:</span>
                                <p>{moment(createdAt).format('ll')}</p>

                                <span>Written by:</span>
                                <p>{author}</p>

                                <span>Artwork by:</span>
                                <p>Square Enix</p>
                            </Information>
                            <Body>
                                <p>{body}</p>
                            </Body>
                        </Main>
                    </Content>
                </Container>
            </Wrapper>
        </View>
    );
}


export default Article;