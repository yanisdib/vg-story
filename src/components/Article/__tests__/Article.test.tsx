import { render } from '@testing-library/react';
import moment from 'moment';
import { MemoryRouter } from 'react-router-dom';

import { Article as IArticle } from '../../../interfaces';

import Article from '../Article';


describe('Article component', () => {
    const currentArticle: IArticle = {
        author: 'Yanis',
        backgroundHex: '#FF5B5B',
        body: 'lorem ipsum',
        frontImage: { src: 'hello.jpg' },
        createdAt: 1000,
        id: '12de',
        permalink: 'this-is-a-test',
        title: 'This is a test',
        topics: [
            {
                createdAt: 0,
                id: '1ac',
                name: 'Test',
                permalink: 'test'
            }
        ]
    };

    it('should render without crashing', () => {
        const { container } = render(
            <Article currentArticle={currentArticle} />,
            { wrapper: MemoryRouter }
        );

        expect(container).toMatchSnapshot();
    });

    it('should render the content of the article', () => {
        const { getByText } = render(
            <Article currentArticle={currentArticle} />,
            { wrapper: MemoryRouter }
        );

        const date: string = moment(currentArticle.createdAt).format('ll');

        expect(getByText('Yanis')).toBeInTheDocument();
        expect(getByText(date)).toBeInTheDocument();
        expect(getByText('Square Enix')).toBeInTheDocument();
        expect(getByText('lorem ipsum')).toBeInTheDocument();
    });
});