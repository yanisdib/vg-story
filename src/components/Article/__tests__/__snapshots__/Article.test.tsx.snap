// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component should render without crashing 1`] = `
<div>
  <div
    class="sc-gsnTZi gqYmWm"
  >
    <div
      class="sc-dkzDqf iXMUXw"
    >
      <div
        class="sc-bczRLJ bJuwle"
      >
        <div
          class="sc-hKMtZM qWfsI"
        >
          <div
            class="sc-eCYdqJ bmXWGA"
          >
            <h5>
              <a
                href="/browse/test"
              >
                Test
              </a>
            </h5>
            <h2>
              This is a test
            </h2>
            <h5>
              Written by Yanis â€” 
            </h5>
          </div>
          <div
            class="sc-jSMfEi bDZnFu"
          >
            <img
              alt="This is a test"
              src="hello.jpg"
              title="This is a test"
            />
          </div>
          <main
            class="sc-gKXOVf fMqBBS"
          >
            <div
              class="sc-iBkjds dWrnOW"
            >
              <span>
                Published:
              </span>
              <p>
                Jan 1, 1970
              </p>
              <span>
                Written by:
              </span>
              <p>
                Yanis
              </p>
              <span>
                Artwork by:
              </span>
              <p>
                Square Enix
              </p>
            </div>
            <div
              class="sc-ftvSup daMggH"
            >
              <p>
                lorem ipsum
              </p>
            </div>
          </main>
        </div>
      </div>
    </div>
  </div>
</div>
`;
